sdk/python/kfp_tekton/compiler/_op_to_template.py

```diff
@@ -0,0 +1,183 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kfp.compiler._k8s_helper import convert_k8s_obj_to_json
+from kfp.compiler._op_to_template import _process_obj
+from kfp import dsl
+from kfp.dsl._container_op import BaseOp
+
+
+def _process_base_ops(op: BaseOp):
+    """Recursively go through the attrs listed in `attrs_with_pipelineparams`
+    and sanitize and replace pipeline params with template var string.
+
+    Returns a processed `BaseOp`.
+
+    NOTE this is an in-place update to `BaseOp`'s attributes (i.e. the ones
+    specified in `attrs_with_pipelineparams`, all `PipelineParam` are replaced
+    with the corresponding template variable strings).
+
+    Args:
+        op {BaseOp}: class that inherits from BaseOp
+
+    Returns:
+        BaseOp
+    """
+
+    # map param's (unsanitized pattern or serialized str pattern) -> input param var str
+    map_to_tmpl_var = {
+        (param.pattern or str(param)): '$(inputs.params.%s)' % param.full_name  # Tekton change
+        for param in op.inputs
+    }
+
+    # process all attr with pipelineParams except inputs and outputs parameters
+    for key in op.attrs_with_pipelineparams:
+        setattr(op, key, _process_obj(getattr(op, key), map_to_tmpl_var))
+
+    return op
+
+
+def _op_to_template(op: BaseOp):
+    """Generate template given an operator inherited from BaseOp."""
+    # NOTE in-place update to BaseOp
+    # replace all PipelineParams with template var strings
+    processed_op = _process_base_ops(op)
+    if isinstance(op, dsl.ContainerOp):
+        # default output artifacts
+        # output_artifact_paths = OrderedDict(op.output_artifact_paths)
+        # print(op.output_artifact_paths)
+        # # This should have been as easy as output_artifact_paths.update(op.file_outputs), but the _outputs_to_json function changes the output names and we must do the same here, so that the names are the same
+        # output_artifact_paths.update(sorted(((param.full_name, processed_op.file_outputs[param.name]) for param in processed_op.outputs.values()), key=lambda x: x[0]))
+
+        # output_artifacts = [
+        #      convert_k8s_obj_to_json(
+        #          ArtifactLocation.create_artifact_for_s3(
+        #              op.artifact_location,
+        #              name=name,
+        #              path=path,
+        #              key='runs/{{workflow.uid}}/{{pod.name}}/' + name + '.tgz'))
+        #     for name, path in output_artifact_paths.items()
+        # ]
+
+        # workflow template
+        container = convert_k8s_obj_to_json(
+            processed_op.container
+        )
+
+        step = {'name': processed_op.name}
+        step.update(container)
+
+        # template.update(container)
```

Comment:
I would add an explanation on top that these next set of functionalities are commented out, and we intend to implement them as the project evolves

Reply:
That would be a short-lived comment as I am changing them currently for my next PR, generating `tasks` instead of `steps`

Reply:
added